rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+straight-forward to add flake references to Mercurial repositories
+since Nix already has Mercurial support.
+
+## Flake registries
+
+Flake registries map symbolic flake identifiers (e.g. `nixpkgs`) to
+"direct" flake references (i.e. any type of flake reference that's not
+an indirection). This is a convenience to users, allowing them to do
+
+    nix run nixpkgs:hello
+
+rather than
+
+    nix run github:NixOS/nixpkgs:hello
+
+There are multiple registries:
+
+* The global registry
+  https://raw.githubusercontent.com/NixOS/flake-registry/master/flake-registry.json. (This
+  location can be overriden via the `flake-registry` option.) Nix
+  automatically fetches this registry periodically. The check interval
+  is determined by the `tarball-ttl` option.
+
+* The local registry `~/.config/nix/registry.json`.
+
+* Registry entries specified on the command line.
+
+A registry is a JSON file that looks like this:
+
+```
+{
+    "version": 1,
+    "flakes": {
+        "nixpkgs": {
+            "uri": "github:NixOS/nixpkgs"
+        },
+        ...
+    }
+}
+```
+
+With this registry, flake references resolve as follows:
+
+* `nixpkgs` -> `github:NixOS/nixpkgs`
+* `nixpkgs/release-19.03` -> `github:NixOS/nixpkgs/release-19.03`
+* `nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c` -> `github:NixOS/nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c`
+
+The registries are searched in reverse order. Thus the local registry
+overrides the global registry, and the command line takes precedence
+over the local and global registries.
+
+## `nix` command line interface
+
+I propose to make flakes the primary way to specify packages in the
+`nix` command line interface. Note that the `nix` UI is still marked
+as experimental, so we have some freedom to make incompatible
+changes. The legacy commands (`nix-build`, `nix-shell`, `nix-env` and
+`nix-instantiate`) should not be changed to avoid breakage.
+
+Most `nix` subcommands work on a list of arguments called
+"installables" for lack of a better word. For example,
+
+    # nix run nixpkgs:hello dwarffs:dwarffs
+
+takes two flake-based installables. The general form is:
+
+    <flake-ref>(:<attr-path>)?
+
+Examples of installables:
+
+* `nixpkgs:packages.hello`
+* `nixpkgs:hello` - short for `nixpkgs:packages.hello`
+* `nixpkgs/release-19.03:hello` - overrides the Git branch to use
+* `github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082:hello` -
+  specifies the exact Git revision to use
+* `dwarffs` - short for `dwarffs:defaultPackage`
+* `nix:hydraJobs.build.x86_64-darwin`
+* `.:hydraJobs.build.x86_64-darwin` - refers to the flake in the
+  current directory (which can be a dirty Git tree)
+* `.` - short for `.:defaultPackage`
+
+If no argument is given, the default is `.`; thus,
+
+    # nix build
+
+is equivalent to
+
+    # nix build .:defaultPackage
+
+For backwards compatibility, it's possible to use non-flake Nix
+expressions using `-f`, e.g. `nix build -f foo.nix foo.bar`.
```

Comment:
-1 on putting backwards compatible cruft in a tool that we still have the freedom to change! AFAIK, the only thing the current `nix build` can do which `nix-build` can't is display a fancy progress bar, so IMHO `nix-build` should be preferred for non-flake expressions.

Reply:
Is there a consensus on deprecating or continuing the support of the "nix-*" commands? I prefer using the `nix` command because its arguments are consistent and the UX matches other tools such as git and cargo.

If we take your suggestion, should the `-f` argument be removed from other `nix` commands, such as `nix eval`? I personally think doing so would reduce `nix`'s usability as a general-purpose build tool rather than as a flake builder.

Reply:
> Is there a consensus on deprecating or continuing the support of the "nix-*" commands? I prefer using the `nix` command because its arguments are consistent and the UX matches other tools such as git and cargo.

The `nix` command is not even stable yet, so nobody should use them for scripting. We can think about starting deprecation once we give any stability guarantees.

Reply:
I don't think we should remove `-f` because it's unrealistic to think that everything will be converted to flakes right away, and there are non-legacy uses of `-f` (e.g. `nix eval -f ...`).